backend/src/apiserver/model/pipeline_version.go

```diff
@@ -0,0 +1,82 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"fmt"
+)
+
+// PipelineVersionStatus a label for the status of the Pipeline.
+// This is intend to make pipeline creation and deletion atomic.
+type PipelineVersionStatus string
+
+const (
+	PipelineVersionCreating PipelineVersionStatus = "CREATING"
+	PipelineVersionReady    PipelineVersionStatus = "READY"
+	PipelineVersionDeleting PipelineVersionStatus = "DELETING"
+)
+
+type PipelineVersion struct {
+	UUID           string `gorm:"column:UUID; not null; primary_key"`
+	CreatedAtInSec int64  `gorm:"column:CreatedAtInSec; not null; index"`
+	Name           string `gorm:"column:Name; not null; unique"`
+	// Set size to 65535 so it will be stored as longtext.
+	// https://dev.mysql.com/doc/refman/8.0/en/column-count-limit.html
+	Parameters string `gorm:"column:Parameters; not null; size:65535"`
+	// PipelineVersion belongs to Pipeline. If a pipeline with a specific UUID
+	// is deleted from Pipeline table, all this pipeline's versions will be
+	// deleted from PipelineVersion table.
+	PipelineId string                `gorm:"column:PipelineId; not null;index;"`
+	Status     PipelineVersionStatus `gorm:"column:Status; not null"`
+	// Code source refers to the pipeline version's definition.
+	// We allow multiple code source references for a single pipeline version.
```

Comment:
url for package storage and url for git snapshot?

Reply:
package storage is not code source. A package is compiled from a source code. Is it a safe assumption to only support single source of truth?

Reply:
Let me try to summarize: (1) in storage/model, we have a single code source url (2) in api, we'll have to have a package url and a code source url in PipelineVersion message, because the package url we need to read file from for CreatePipelineVersionRequest. But the package url is not stored in db, so not in storage/model definition...